core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -301,50 +412,171 @@ private PrivateMetadata generateMissingInternalMetadata() {
       return builder.build();
    }
 
-   private void migrateFromV12_0()  throws Exception {
+   private void migrateFromComposedSegmentedLoadWriteStore(boolean removeOnly) throws IOException {
+      PERSISTENCE.startMigratingPersistenceData(cacheName());
+      File newFile = new File(file.getParentFile(), cacheName() + "_new.dat");
+      try {
+         if (SecurityActions.fileExists(newFile)) {
+            if (log.isTraceEnabled()) log.tracef("Overwriting temporary migration file %s", newFile);
+            // Delete file to overwrite permissions as well
+            SecurityActions.deleteFile(newFile);
+         }
+
+         try (FileChannel newChannel = createNewFile(newFile)) {
+            if (!removeOnly) {
+               copyEntriesFromOldSegmentFiles(newChannel);
+            }
+         }
+
+         // Move the new file to the final name
+         SecurityActions.moveFile(newFile.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+         // Reopen the file
+         channel = SecurityActions.openFileChannel(file);
+
+         // Remove the composed segment files
+         removeComposedSegmentedLoadWriteStoreFiles();
+
+         PERSISTENCE.persistedDataSuccessfulMigrated(cacheName());
+      } catch (PersistenceException e) {
+         throw e;
+      } catch (Exception e) {
+         throw PERSISTENCE.persistedDataMigrationFailed(cacheName(), e);
+      }
+   }
+
+   private void copyEntriesFromOldSegmentFiles(FileChannel newChannel) throws Exception {
+      int numSegments = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
```

Comment:
I wonder if we should throw an exception instead if the number of segments don't match the number of files. We are sorta covering the case if we have more segments than before but if we configure less it will miss some data.

Reply:
I thought about it as well, but it's going to get more complicated when the cache is clustered and each node only has a subset of the segments: the migration might succeed or fail depending on exactly which segments were local. I thought it would be simpler if we just recommend keeping the number of segments unchanged.

Does SIFS check the number of segments when migrating data from SFS? Perhaps I could use the same checks.

Reply:
> I thought it would be simpler if we just recommend keeping the number of segments unchanged.

Oh I think that is exactly what we should do. Just reading the code however made me think of supporting other configurations. The more I am reading it again that is all this is doing as well. The checks for file not existing just confused me at first, but that can happen for DIST. So we can close this out :)

> Does SIFS check the number of segments when migrating data from SFS? Perhaps I could use the same checks.

This only works with the same number of segments.